### PR TITLE
nonempty grid rows

### LINKED ISSUE #1843: Use `NonEmpty` and separate `EmptyGrid` constructor for `Grid`
With a grid definition like `newtype Grid = Grid [[a]]`, there is more than one way to represent an "empty grid":
* `Grid []`
* `Grid [[]]`
* `Grid [[], [], [], [], [], ...]`

If instead we use `NonEmpty` for both rows and columns, then there can be exactly one "normalized" representation for an empty grid:
```
data Grid a = EmptyGrid | Grid (NonEmpty (NonEmpty a))
```

### PR DESCRIPTION
**Pull request by kostmo**
Closes #1843

Towards #1981

### CODE REVIEW

File: src/swarm-topography/Swarm/Game/Scenario/Topography/Area.hs [hunk 1]
```diff
@@ -94,8 +112,13 @@ getAreaDimensions cellGrid =
 computeArea :: AreaDimensions -> Int32
 computeArea (AreaDimensions w h) = w * h
 
+-- |
+-- Warning: size must be nonzero,
```
Comment (byorgey) [inline src/swarm-topography/Swarm/Game/Scenario/Topography/Area.hs:116]: This function seems unnecessarily partial.  Why not have a special case to first check whether `w` or `h` is zero, and if so return `EmptyGrid`?

File: src/swarm-scenario/Swarm/Game/Scenario/Topography/WorldDescription.hs [hunk 2]
```diff
@@ -18,7 +18,7 @@ import Swarm.Game.Entity
 import Swarm.Game.Land
 import Swarm.Game.Location
 import Swarm.Game.Scenario.RobotLookup
-import Swarm.Game.Scenario.Topography.Area (emptyGrid)
+import Swarm.Game.Scenario.Topography.Area (Grid (EmptyGrid))
```
Comment (byorgey) [inline src/swarm-scenario/Swarm/Game/Scenario/Topography/WorldDescription.hs:21]: It seems like a bit of a code smell to import a constructor of `Grid` just to be able to make an empty grid, especially if the eventual goal is to make a smart constructor.  I think the constructors of `Grid` should not be exported at all.  That way this module (and others like it) won't have to be updated if the constructors ever change again. We could either (1) continue to export `emptyGrid` as before, or (2) (slightly fancier) make an instance of `AsEmpty` and then import `pattern Empty` from `Control.Lens.Empty`.
```
instance AsEmpty Grid where
  _Empty = prism' (const EmptyGrid) (\case {EmptyGrid -> Just (); _ -> Nothing})
```

File: src/swarm-topography/Swarm/Game/Scenario/Topography/Rasterize.hs [hunk 3]
```diff
@@ -5,13 +5,15 @@
 module Swarm.Game.Scenario.Topography.Rasterize where
 
 import Codec.Picture
+import Data.List.NonEmpty qualified as NE
 import Data.Vector qualified as V
 import Swarm.Game.Scenario.Topography.Area
 
 -- | Converts linked lists to vectors to facilitate
 -- random access when assembling the image
 gridToVec :: Grid a -> V.Vector (V.Vector a)
```
Comment (byorgey) [inline src/swarm-topography/Swarm/Game/Scenario/Topography/Rasterize.hs:14]: I think this function should move to `Swarm.Game.Scenario.Topography.Area` (especially if we stop exporting the constructors of `Grid`).

File: src/swarm-topography/Swarm/Game/Scenario/Topography/Structure/Overlay.hs [hunk 4]
```diff
@@ -68,14 +68,14 @@ computeMergedArea (OverlayPair pg1 pg2) =
 zipGridRows ::
   Alternative f =>
   AreaDimensions ->
-  OverlayPair (Grid (f a)) ->
+  OverlayPair [[f a]] ->
```
Comment (byorgey) [inline src/swarm-topography/Swarm/Game/Scenario/Topography/Structure/Overlay.hs:71]: What is the rationale for changing `Grid (f a)` to `[[f a]]` here?  It seems like a step backward in terms of generality.

File: src/swarm-topography/Swarm/Game/Scenario/Topography/Structure/Overlay.hs [hunk 5]
```diff
@@ -124,11 +124,11 @@ padSouthwest ::
   Alternative f =>
   V2 Int32 ->
   OverlayPair (Grid (f a)) ->
-  OverlayPair (Grid (f a))
+  OverlayPair [[f a]]
```
Comment (byorgey) [inline src/swarm-topography/Swarm/Game/Scenario/Topography/Structure/Overlay.hs:127]: Same question here.

**Review by byorgey — APPROVED** [review]

File: src/swarm-topography/Swarm/Game/Scenario/Topography/Structure/Overlay.hs (hunk 4 above)
Comment (kostmo) [reply src/swarm-topography/Swarm/Game/Scenario/Topography/Structure/Overlay.hs:72]: Yeah, I had attempted `Grid (f a)` at first, but encountered a bug in `padSouthwest` that was causing structures to get blanked out.  I didn't have time to fully investigate it.  Created #2004 to follow up.

File: src/swarm-scenario/Swarm/Game/Scenario/Topography/WorldDescription.hs (hunk 2 above)
Comment (kostmo) [reply src/swarm-scenario/Swarm/Game/Scenario/Topography/WorldDescription.hs]: For now there are still a number of places where the `Grid` constructor is used.  Hiding exports of the constructors will be a prerequisite for completing #1981.

**Pull request merged**